packages/snaps-sdk/src/internals/structs.test.ts

```diff
@@ -62,7 +75,52 @@ describe('typedUnion', () => {
     const result = validate(Text({}), nestedUnionStruct);
 
     expect(result[0]?.message).toBe(
-      'At path: props.children -- Expected the value to satisfy a union of `union | array`, but received: undefined',
+      'At path: props.children -- Expected type to be one of: "Bold", "Italic", "Link", "Icon", but received: undefined',
+    );
+  });
+
+  it('validates refined elements', () => {
+    const refinedUnionStruct = typedUnion([BoxStruct, FooterStruct]);
+    const result = validate(
+      Footer({ children: Button({ children: Icon({ name: 'wallet' }) }) }),
+      refinedUnionStruct,
+    );
+
+    expect(result[0]?.message).toBe(
+      'At path: props.children -- Footer buttons may only contain text.',
+    );
+  });
+
+  it('supports coercion', () => {
+    const coercedStruct = typedUnion([
+      BoxStruct,
+      defaulted(object({ type: literal('Custom'), key: string() }), {
+        type: 'Custom',
+        key: 'foo',
+      }),
+    ]);
+
+    const result = create(
+      { type: 'Custom' },
+      coercedStruct,
+    ) as GenericSnapElement;
```

Comment:
Why does this need a type cast?

Reply:
It doesn't, must be from some other iteration, will remove.